### PR TITLE
NIFI-3838: Initial implementation of RecordPath and UpdateRecord proc…

### PR DESCRIPTION
**Pull request by markap14**
…essor

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1459,6 +1459,11 @@
             </dependency>
             <dependency>
                 <groupId>org.apache.nifi</groupId>
+                <artifactId>nifi-record-path</artifactId>
+                <version>1.2.0-SNAPSHOT</version>
```
Comment (joewitt) [inline pom.xml]: @markap14  this should be 1.3.0-SNAPSHOT

File: nifi-commons/nifi-record-path/pom.xml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+      See the License for the specific language governing permissions and
+      limitations under the License.
+    -->
+
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>org.apache.nifi</groupId>
+    <artifactId>nifi-commons</artifactId>
+    <version>1.2.0-SNAPSHOT</version>
```
Comment (joewitt) [inline nifi-commons/nifi-record-path/pom.xml]: @markap14  this should be 1.3.0-SNAPSHOT

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml [hunk 3]
```diff
@@ -233,6 +237,12 @@
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-record-serialization-services</artifactId>
+            <version>1.2.0-SNAPSHOT</version>
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/pom.xml]: @markap14  this should be 1.3.0-SNAPSHOT

**Comment by markap14** [comment]
@joewitt good catch. Addressed the versioning info.

File: nifi-commons/nifi-record-path/src/main/java/org/apache/nifi/record/path/ArrayIndexFieldValue.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    public ArrayIndexFieldValue(final Object value, final RecordField field, final FieldValue parent, final int index) {
+        super(value, field, validateParent(parent));
+        this.index = index;
+    }
+
+    private static FieldValue validateParent(final FieldValue parent) {
+        Objects.requireNonNull(parent, "Cannot create an ArrayIndexFieldValue without a parent");
+        if (RecordFieldType.ARRAY != parent.getField().getDataType().getFieldType()) {
+            // TODO: Add Unit test that does something like //myArray[../name = 'myName'][0] and ensure that
```
Comment (mattyb149) [inline nifi-commons/nifi-record-path/src/main/java/org/apache/nifi/record/path/ArrayIndexFieldValue.java]: Check your TODOs please to see which should stay and which can go :)

File: nifi-docs/src/main/asciidoc/record-path-guide.adoc [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+ 
+We can now reference the first element in the `addresses` array by using the RecordPath `/addresses[0]`. We can access the second element using the RecordPath `/addresses[1]`.
+There may be times, though, that we don't know how many elements will exist in the array. So we can use negative indices to count backward from the end of the array. For example,
+we can access the last element as `/addresses[-1]` or the next-to-last element as `/addresses[-2]`. If we want to reference several elements, we can use a comma-separated list of
+elements, such as `/addresses[0, 1, 2, 3]`. Or, to access elements 0 through 8, we can use the `range` operator (`..`), as in `/addresses[0..8]`. We can also mix these, and reference
+all elements by using the syntax `/addresses[0..-1]` or even `/addresses[0, 1, 4, 6..-1]`. Of course, not all of the indices referenced here will match on the Record above, because
+the `addresses` array has only 2 elements. The indices that do not match will simply be skipped.
+
+
+[[maps]
```
Comment (mattyb149) [inline nifi-docs/src/main/asciidoc/record-path-guide.adoc]: Missing a closing brace here to make a Section

File: nifi-docs/src/main/asciidoc/record-path-guide.adoc [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+[[predicates]]
+=== Predicates
+
+Thus far, we have discussed two different types of filters. Each of them allows us to select one or more elements out from a field that allows for many values.
+Often times, though, we need to apply a filter that allows us to restrict which Record fields are selected. For example, what if we want to select the `zip` field but
+only for an `address` field where the state is not New York? The above examples do not give us any way to do this.
+
+RecordPath provides the user the ability to specify a Predicate. A Predicate is simply a filter that can be applied to a field in order to determine whether or not the
+field should be included in the results. Like other filters, a Predicate is specified within square brackets. The syntax of the Predicate is
+`<Relative RecordPath> <Operator> <Expression>`. The `Relative RecordPath` works just like any other RecordPath but must start with a `.` instead of a slash and references
```
Comment (mattyb149) [inline nifi-docs/src/main/asciidoc/record-path-guide.adoc]: Maybe an example of the double-dot relative path usage too? Searching for the double dot on this page only brings you to the Range section

**Comment by markap14** [comment]
@mattyb149 I just pushed a new commit that should address the issue that you noticed here with the differing schemas. I added unit tests to verify as well. I just noticed your comment about the documentation and will push another commit momentarily to address that.

**Comment by markap14** [comment]
@mattyb149 updated docs. Thanks!

**Comment by markap14** [comment]
@mattyb149 I also ran into an issue when comparing Integer values to Long values so I pushed one more commit to handle that case.

**Comment by mattyb149** [comment]
+1 LGTM, ran full build with unit tests, tried various scenarios in UpdateRecord (relative path predicates, same-name-different-type, etc.)  Merging to master

**Pull request closed without merging**